ext/afform/core/Civi/Api4/Action/Afform/Submit.php

```diff
@@ -72,26 +72,33 @@ public static function processGenericEntity(AfformSubmitEvent $event) {
         $api4 = $event->formDataModel->getSecureApi4($entityName);
         foreach ($records as $record) {
           $saved = $api4($entityType, 'save', ['records' => [$record['fields']]])->first();
-          self::saveJoins($api4, $entityType, $saved['id'], $record['joins'] ?? []);
+          self::saveJoins($entityType, $saved['id'], $record['joins'] ?? []);
         }
       }
       unset($event->entityValues[$entityType]);
     }
   }
 
-  protected static function saveJoins($api4, $mainEntityName, $entityId, $joins) {
+  protected static function saveJoins($mainEntityName, $entityId, $joins) {
     foreach ($joins as $joinEntityName => $join) {
       $values = self::filterEmptyJoins($joinEntityName, $join);
-      // FIXME: Replace/delete should only be done to known contacts
+      // TODO: REPLACE works for creating or updating contacts, but different logic would be needed if
+      // the contact was being auto-updated via a dedupe rule; in that case we would not want to
+      // delete any existing records.
       if ($values) {
-        $api4($joinEntityName, 'replace', [
+        civicrm_api4($joinEntityName, 'replace', [
+          // Disable permission checks because the main entity has already been vetted
+          'checkPermissions' => FALSE,
```

Comment:
So this seems to be the guts of the pr - what limits it to only address etc entities and not others?

Reply:
This will only fire for `af-join` entitites which are like "sub-entities" of a Contact record (aside: in theory it could be possible to have a sub-entity of something else but so far it's just Contacts). And it only fires *after* the Contact has been saved, which would have thrown an exception if authentication failed. So at this point we already know that saving the Contact is allowed.

Reply:
@colemanw ok - but how do we ensure that when we extend the entities we don't accidentally expose more inappropriate sub entities - how are they defined? Presumably pro-actively?

Reply:
Also what if you are trying to update a joined case say from an activity but you don't have case permissions how would this work?

Reply:
Yes they have to be explicitly defined in code, e.g. [`afjoinAddressDefault.aff.json`](https://github.com/colemanw/civicrm-core/blob/9e58ceedbeba33102e5a02cee135966833337158/ext/afform/core/ang/afjoinAddressDefault.aff.json). And then they have to be explicitly added to the form by the form author.

Reply:
@seamuslee001 `af-join`s really are *sub*-entities. Other than email, phone, address, website, im, and multi-record custom fields, I don't expect there to be many other entities that follow this pattern.

Activities and Cases are both full entities in their own right. We wouldn't use the `af-join` pattern with them. We'd probably use an EntityRef to link them, although we'd have to do something about bridging them through the `ActivityContact` table. But that's a separate issue to this.

Reply:
@colemanw I think that deals with my concerns on a technical level but I'm still nervous on a docs level - ie the concept of sub-entities & implications should be documented. I think the urgency of having a home for afform documentation is increasing (@MikeyMJCO  @totten @seamuslee001 )

Reply:
I've added https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/913 to start to establish a docs location. It is very minimal but hopefully we could put something about how sub-entities are declared in there